fx quant: clarify state in Quantizer object

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43928 fx quant: add docblocks to _find_matches and _find_quants
* **#43927 fx quant: clarify state in Quantizer object**

Summary:

Adds uninitialized placeholders for various state
used throughout the Quantizer object, with documentation
on what they are. No logic change.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23439473](https://our.internmc.facebook.com/intern/diff/D23439473)

vkuzo committed:
fx quant: clarify state in Quantizer object

Summary:

Adds uninitialized placeholders for various state
used throughout the Quantizer object, with documentation
on what they are. No logic change.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
nit on "fx quant: clarify state in Quantizer object"


Summary:

Adds uninitialized placeholders for various state
used throughout the Quantizer object, with documentation
on what they are. No logic change.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23439473](https://our.internmc.facebook.com/intern/diff/D23439473)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fx/quantize.py:
```diff
@@ -599,6 +599,19 @@ def __init__(self):
         # mapping from matched node to activation_post_process
         # must be filled before convert
         self.activation_post_process_map = None
+        # mapping from node name to qconfig that should be used for that node
```
please hold off this one, this is refactored in https://github.com/pytorch/pytorch/pull/43901 and we no longer need to initialize them in __init__

vkuzo replied:
sure, sounds good

vkuzo replied:
I read through the relevant sections of https://github.com/pytorch/pytorch/pull/43901/files, not sure I understand your suggestion.  Can you give some more context?

this PR is not adding or changing any functionality, it just improves documentation.  It's not intuitive to have a Python class where important state is not declared in `__init__`.

jerryzh168 replied:
I see, if this is just used as declaration then it might be OK, but we'll need to make sure in convert we distinguish the declared and defined states correctly

vkuzo replied:
sg.  I think for people who are not the original code authors but who need to understand/modify the code, glancing at `__init__` to see what the important state is for a class is pretty common.  This should make it easier for others to quickly understand the high level without having to read every LOC.

jerryzh168 commented on torch/quantization/fx/quantize.py:
```diff
@@ -599,6 +599,19 @@ def __init__(self):
+        #   'linear.weight_fake_quant': PerChannelMinMaxObserver(...),
+        # }
+        self.modules = None
+        # mapping from observer node name to observer instance
+        self.activation_post_process_map = None
```
also this is defined before in L601

vkuzo committed:
Update on "fx quant: clarify state in Quantizer object"


Summary:

Adds uninitialized placeholders for various state
used throughout the Quantizer object, with documentation
on what they are. No logic change.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23439473](https://our.internmc.facebook.com/intern/diff/D23439473)

[ghstack-poisoned]

vkuzo committed:
Update on "fx quant: clarify state in Quantizer object"


Summary:

Adds uninitialized placeholders for various state
used throughout the Quantizer object, with documentation
on what they are. No logic change.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23439473](https://our.internmc.facebook.com/intern/diff/D23439473)

[ghstack-poisoned]

jerryzh168 reviewed (changes requested):
please change the checks accordingly: https://github.com/pytorch/pytorch/blob/master/torch/quantization/fx/quantize.py#L253

vkuzo commented:
> please change the checks accordingly: https://github.com/pytorch/pytorch/blob/master/torch/quantization/fx/quantize.py#L253

Would you mind clarifying? It looks like the current logic should work as is because `save_state` renames `self.patterns` to `self._patterns`, and so on.

jerryzh168 commented:
> > please change the checks accordingly: https://github.com/pytorch/pytorch/blob/master/torch/quantization/fx/quantize.py#L253
> 
> Would you mind clarifying? It looks like the current logic should work as is because `save_state` renames `self.patterns` to `self._patterns`, and so on.

oh, i see. what I meant is that now these attributes are initialized to None, so we have two invalid cases, one is that observed module has no such attributes which is covered in the check, the other is the attributes are None.

but I guess from the code it probably won't happen that the attributes are None. unless user calls save_state explicitly, so probably we'll need to change 'save_state' to '_save_state' to make sure it's not called by user, will accept for now.

vkuzo commented:
> > > please change the checks accordingly: https://github.com/pytorch/pytorch/blob/master/torch/quantization/fx/quantize.py#L253
> > 
> > 
> > Would you mind clarifying? It looks like the current logic should work as is because `save_state` renames `self.patterns` to `self._patterns`, and so on.
> 
> oh, i see. what I meant is that now these attributes are initialized to None, so we have two invalid cases, one is that observed module has no such attributes which is covered in the check, the other is the attributes are None.
> 
> but I guess from the code it probably won't happen that the attributes are None. unless user calls save_state explicitly, so probably we'll need to change 'save_state' to '_save_state' to make sure it's not called by user, will accept for now.

makes sense.  In a future PR, perhaps we can either regenerate the state dynamically (as we chatted on a previous PR) or use original var names to for saving it (in which case the checks would need to be changed).

vkuzo committed:
Update on "fx quant: clarify state in Quantizer object"


Summary:

Adds uninitialized placeholders for various state
used throughout the Quantizer object, with documentation
on what they are. No logic change.

Test Plan:

```
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFx
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23439473](https://our.internmc.facebook.com/intern/diff/D23439473)

[ghstack-poisoned]
